### ISSUE TITLE
Running instance on remote server without docker and sudo

### ISSUE THREAD

**Issue by zyzhang1130** [huawei-noah/SMARTS#1635]
**High Level Description**
I try to run an instance on a remote linux server without docker and sudo, since it is not allowed (due to security concern). 

**Desired SMARTS version**
0.6.1

**Operating System**
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"

**Problems**
When I did test with `make sanity-test`, it showed the following error: 
```
make sanity-test
scl scenario build --clean scenarios/sumo/loop
The `scl waymo` command is unavailable. To enable, pip install the missing dependencies.
pip install pathos==0.2.8 tabulate>=0.8.10 waymo-open-dataset-tf-2-4-0
build-scenario scenarios/sumo/loop
Traceback (most recent call last):
  File "tmpw6hzx7xd.py", line 42, in <module>
    gen_scenario(
  File "/home/zeyu/SMARTS/smarts/sstudio/genscenario.py", line 67, in gen_scenario
    gen_traffic(
  File "/home/zeyu/SMARTS/smarts/sstudio/genscenario.py", line 162, in gen_traffic
    saved_path = generator.plan_and_save(traffic, name, output_dir, seed=seed)
  File "/home/zeyu/SMARTS/smarts/sstudio/generators.py", line 186, in plan_and_save
    duarouter(
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/sh.py", line 1524, in __call__
    return RunningCommand(cmd, call_args, stdin, stdout, stderr)
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/sh.py", line 788, in __init__
    self.wait()
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/sh.py", line 845, in wait
    self.handle_command_exit_code(exit_code)
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/sh.py", line 869, in handle_command_exit_code
    raise exc
sh.ErrorReturnCode_127: 

  RAN: /home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/sumo/bin/duarouter --unsorted-input --net-file=./map.net.xml --route-files=/tmp/tmpoygb4znh/trips.trips.xml --output-file=./traffic/basic.rou.xml --seed=42 --no-step-log --repair --error-log=/home/zeyu/.smarts/_duarouter_routing/traffic/basic.log

  STDOUT:


  STDERR:
/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/sumo/bin/duarouter: error while loading shared libraries: libXrender.so.1: cannot open shared object file: No such file or directory

Traceback (most recent call last):
  File "/home/zeyu/SMARTS/.venv/bin/scl", line 33, in <module>
    sys.exit(load_entry_point('smarts', 'console_scripts', 'scl')())
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/zeyu/SMARTS/.venv/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/zeyu/SMARTS/cli/studio.py", line 65, in build_scenario
    build_single_scenario(clean, allow_offset_map, scenario, seed, click.echo)
  File "/home/zeyu/SMARTS/smarts/sstudio/build_scenario.py", line 53, in build_single_scenario
    subprocess.check_call(
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/zeyu/SMARTS/.venv/bin/python3.8', 'tmpw6hzx7xd.py']' returned non-zero exit status 1.
make: *** [Makefile:82: build-sanity-scenarios] Error 1
```
I think it has to do with the fact that libXrender is not installed on the server. Is it possible to run the instance without libXrender, since my intention is to render the visuals on my local machine (I read from the SMARTS paper that it is possible but I'm also not sure how to do it since it seems some extra setup is required)? Also installing libXrender requires sudo which I do not have access to.

Many thanks.

UPDATE: Now I am running into another issue. When the training started, I'll keep getting this error message repeatedly after a while:
```
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ python3.8 run.py

Torch cuda is available:  True 


Logdir: /home/z/zhan1130/SMARTS/competition/track1/train/logs/2022_09_22_23_31_36 

Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_c/. ...
ERROR:websocket:[Errno 111] Connection refused - goodbye
WARNING:Client:Connection to Envision failed with: [Errno 111] Connection refused.  Try using `--headless` if not using Envision server.
/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/gym/logger.py:34: UserWarning: WARN: Box bound precision lowered by casting to float32
  warnings.warn(colorize("%s: %s" % ("WARN", msg % args), "yellow"))
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_c/. ...
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
ERROR:websocket:[Errno 111] Connection refused - goodbye
```
I'm not sure if it is related to sending visuals from the remote server to my local machine, but I haven't figured out a way to resolve it..

**Comment by Adaickalavan** [comment]
Hi @zyzhang1130,

1) Are you trying to run SMARTS directly on a remote server without Docker and without sudo?

2) Could you provide the exact sequence of commands executed which resulted in the message 
    ```txt
    ERROR:websocket:[Errno 111] Connection refused - goodbye
    WARNING:Client:Connection to Envision failed with: [Errno 111] Connection refused.  Try using `--headless` if not using Envision server.
    ```

3) From the message, it appears that the Envision server might not have started. Did you start an Envision server using `scl envision start -s <scenario_dir>` ?

**Comment by zyzhang1130** [comment]
> Hi @zyzhang1130,
> 
>     1. Are you trying to run SMARTS directly on a remote server without Docker and without sudo?
> 
>     2. Could you provide the exact sequence of commands executed which resulted in the message
>        ```
>        ERROR:websocket:[Errno 111] Connection refused - goodbye
>        WARNING:Client:Connection to Envision failed with: [Errno 111] Connection refused.  Try using `--headless` if not using Envision server.
>        ```
> 
>     3. From the message, it appears that the Envision server might not have started. Did you start an Envision server using `scl envision start -s <scenario_dir>` ?

1. Yes. I can neither using docker nor sudo.
2.  I followed the instructions given here except skipping step 4 as I was not using docker: https://github.com/huawei-noah/SMARTS/issues/1624#issuecomment-1242810034

```
(base) zhan1130@xgpb0:~$ cd /SMARTS/competition/track1/train
bash: cd: /SMARTS/competition/track1/train: No such file or directory
(base) zhan1130@xgpb0:~$ cd SMARTS/competition/track1/train
(base) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ conda activate smarts
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ scl envision start -s /opt/.venv/lib/python3.8/site-packages/smarts/scenarios/ &
[1] 5804
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ python3.8 run.py --mode=evaluate --model=/src/competition/track1/submission/best_model.zip 

Torch cuda is available:  True 


Logdir: /home/z/zhan1130/SMARTS/competition/track1/train/logs/2022_09_24_17_36_15 


Model: /src/competition/track1/submission/best_model.zip 

Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_c/. ...
/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/gym/logger.py:34: UserWarning: WARN: Box bound precision lowered by casting to float32
  warnings.warn(colorize("%s: %s" % ("WARN", msg % args), "yellow"))
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_c/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_t/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_t/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/merge/3lane_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/merge/3lane_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_cruise_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_cruise_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_overtake/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_overtake/. ...

Evaluate policy.

== CURRENT SYSTEM INFO ==
OS: Linux-5.4.0-126-generic-x86_64-with-glibc2.17 #142-Ubuntu SMP Fri Aug 26 12:12:57 UTC 2022
Python: 3.8.13
Stable-Baselines3: 1.4.0
PyTorch: 1.12.1+cu113
GPU Enabled: True
Numpy: 1.21.5
Gym: 0.19.0

Traceback (most recent call last):
  File "run.py", line 158, in <module>
    main(args)
  File "run.py", line 70, in main
    run(envs_train=envs_train, envs_eval=envs_eval, config=config)
  File "run.py", line 119, in run
    model = getattr(sb3lib, config["alg"]).load(
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/stable_baselines3/common/base_class.py", line 687, in load
    data, params, pytorch_variables = load_from_zip_file(
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/stable_baselines3/common/save_util.py", line 388, in load_from_zip_file
    load_path = open_path(load_path, "r", verbose=verbose, suffix="zip")
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/stable_baselines3/common/save_util.py", line 232, in open_path_str
    return open_path(pathlib.Path(path), mode, verbose, suffix)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/stable_baselines3/common/save_util.py", line 284, in open_path_pathlib
    return open_path(path, mode, verbose, suffix)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/stable_baselines3/common/save_util.py", line 264, in open_path_pathlib
    raise error
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/stable_baselines3/common/save_util.py", line 256, in open_path_pathlib
    path = path.open("rb")
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/pathlib.py", line 1222, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/pathlib.py", line 1078, in _opener
    return self._accessor.open(self, flags, mode)
FileNotFoundError: [Errno 2] No such file or directory: '/src/competition/track1/submission/best_model.zip.zip'
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='/home/z/zhan1130/SMARTS/competition/track1/train/logs/2022_09_24_17_36_15/monitor.csv' mode='wt' encoding='UTF-8'>
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function SMARTS.__del__ at 0x7f601616baf0>
Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/core/smarts.py", line 856, in __del__
TypeError: 'NoneType' object is not callable
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ 
```
4. yes I executed `scl envision start -s /opt/.venv/lib/python3.8/site-packages/smarts/scenarios/ & ` as it is one step in the instruction given above

**Comment by Adaickalavan** [comment]
Hi @zyzhang1130,

1) The command to start Envision server is
    ```
    $ scl envision start -s <path_to_scenario_files>
    ```
    When running SMARTS inside our given Docker image, the `<path_to_scenario_files>` was `/opt/.venv/lib/python3.8/site-packages/smarts/scenarios/`. In other situations, we need to use the correct path to the scenario files. The scenario files are located inside the SMARTS package installed inside the python virtual enviroment.
    + Where is your python virtual environment located?
    + Generally, the <path_to_scenario_files> would be like `<path to installed python packages>/smarts/scenario/`

2) When running the example code inside our given Docker image, the location of the saved model was `/src/competition/track1/submission/best_model.zip`. In other situations, we need to use the correct path to the saved model. 

    Taking this command as an example, 
    ```bash
    zhan1130@xgpb0:~/SMARTS/competition/track1/train$ python3.8 run.py --mode=evaluate --model=/src/competition/track1/submission/best_model.zip 
    ```
    the saved model is located at `~/SMARTS/competition/track1/submission/best_model.zip`. Please use the full absolute path to the saved model and do not use `~` when supplying the saved model path as an input argument.

**Comment by zyzhang1130** [comment]
> 



> Hi @zyzhang1130,
> 
>     1. The command to start Envision server is
>        ```
>        $ scl envision start -s <path_to_scenario_files>
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        When running SMARTS inside our given Docker image, the `<path_to_scenario_files>` was `/opt/.venv/lib/python3.8/site-packages/smarts/scenarios/`. In other situations, we need to use the correct path to the scenario files. The scenario files are located inside the SMARTS package installed inside the python virtual enviroment.
>        
>        * Where is your python virtual environment located?
>        * Generally, the `<path_to_scenario_files>` would be like `<path to installed python packages>/smarts/scenario/`
> 
>     2. When running the example code inside our given Docker image, the location of the saved model was `/src/competition/track1/submission/best_model.zip`. In other situations, we need to use the correct path to the saved model.
>        Taking this command as an example,
>        ```shell
>        zhan1130@xgpb0:~/SMARTS/competition/track1/train$ python3.8 run.py --mode=evaluate --model=/src/competition/track1/submission/best_model.zip 
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        the saved example model is located at `~/SMARTS/competition/track1/submission/best_model.zip`. Please use the full absolute path to the saved model and do not use `~` when supplying the saved model path as an input argument.

1. actually `scl envision start -s /opt/.venv/lib/python3.8/site-packages/smarts/scenarios/ & ` does bring me to the correct directory:

```
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ scl envision start -s /opt/.venv/lib/python3.8/site-packages/smarts/scenarios/ & 
[2] 268856
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ Traceback (most recent call last):
  File "/home/z/zhan1130/anaconda3/envs/smarts/bin/scl", line 8, in <module>
    sys.exit(scl())
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/cli/envision.py", line 53, in start_server
    run(scenario_dirs=scenarios, max_capacity_mb=max_capacity, port=port)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/envision/server.py", line 524, in run
    app.listen(port)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/tornado/web.py", line 2109, in listen
    server.listen(port, address)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/tornado/tcpserver.py", line 151, in listen
    sockets = bind_sockets(port, address=address)
  File "/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/tornado/netutil.py", line 161, in bind_sockets
    sock.bind(sockaddr)
OSError: [Errno 98] Address already in use
cd /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages
[2]+  Exit 1                  scl envision start -s /opt/.venv/lib/python3.8/site-packages/smarts/scenarios/  (wd: ~/SMARTS/competition/track1/train)
(wd now: ~/anaconda3/envs/smarts/lib/python3.8/site-packages)

(smarts) zhan1130@xgpb0:~/anaconda3/envs/smarts/lib/python3.8/site-packages$ cd smarts
(smarts) zhan1130@xgpb0:~/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts$ ls
contrib  core  env  __init__.py  __pycache__  ros  scenarios  sstudio  waymo  zoo
(smarts) zhan1130@xgpb0:~/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts$ cd scnarios
bash: cd: scnarios: No such file or directory
(smarts) zhan1130@xgpb0:~/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts$ cd scenarios
(smarts) zhan1130@xgpb0:~/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios$ ls
intersection  merge  straight
(smarts) zhan1130@xgpb0:~/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios$ 
```
2. changed to the absolute path of the model and it gave the following error:
```(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ python3.8 run.py --mode=evaluate --model=/home/z/zhan1130/SMARTS/competition/track1/submission/best_model.zip

Torch cuda is available:  True 


Logdir: /home/z/zhan1130/SMARTS/competition/track1/train/logs/2022_09_25_22_50_28 


Model: /home/z/zhan1130/SMARTS/competition/track1/submission/best_model.zip 

Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_c/. ...
/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/gym/logger.py:34: UserWarning: WARN: Box bound precision lowered by casting to float32
  warnings.warn(colorize("%s: %s" % ("WARN", msg % args), "yellow"))
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_c/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_t/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_t/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/merge/3lane_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/merge/3lane_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_cruise_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_cruise_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_overtake/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_overtake/. ...

Evaluate policy.

== CURRENT SYSTEM INFO ==
OS: Linux-5.4.0-126-generic-x86_64-with-glibc2.17 #142-Ubuntu SMP Fri Aug 26 12:12:57 UTC 2022
Python: 3.8.13
Stable-Baselines3: 1.4.0
PyTorch: 1.12.1+cu113
GPU Enabled: True
Numpy: 1.21.5
Gym: 0.19.0

== SAVED MODEL SYSTEM INFO ==
OS: Linux-5.13.0-44-generic-x86_64-with-glibc2.29 #49~20.04.1-Ubuntu SMP Wed May 18 18:44:28 UTC 2022
Python: 3.8.10
Stable-Baselines3: 1.4.0
PyTorch: 1.12.0+cu102
GPU Enabled: True
Numpy: 1.23.1
Gym: 0.19.0

/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/stable_baselines3/common/base_class.py:687: ResourceWarning: unclosed file <_io.BufferedReader name='/home/z/zhan1130/SMARTS/competition/track1/submission/best_model.zip'>
  data, params, pytorch_variables = load_from_zip_file(
ResourceWarning: Enable tracemalloc to get the object allocation traceback

Evaluating env 1_to_2lane_left_turn_c.
 Retrying in 0.05 seconds
Could not connect to TraCI server at localhost:37097 [Errno 111] Connection refused
 Retrying in 0.05 seconds
Could not connect to TraCI server at localhost:37097 [Errno 111] Connection refused
 Retrying in 0.05 seconds
Could not connect to TraCI server at localhost:37097 [Errno 111] Connection refused
 Retrying in 0.05 seconds
Could not connect to TraCI server at localhost:37097 [Errno 111] Connection refused
 Retrying in 0.05 seconds
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Evaluating env 1_to_2lane_left_turn_t.
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Evaluating env 3lane_merge_single_agent.
 Retrying in 0.05 seconds
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Evaluating env 3lane_cruise_single_agent.
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Evaluating env 3lane_overtake.
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Finished evaluating.

WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ 
```

**Comment by Adaickalavan** [comment]
Originally, there were two issues, namely,
1. Incorect path to the scenario files.
2. Incorrect path to the saved model. 

Now, do the following:
1. Kill any running python processes and Envision servers.
2. On a new terminal, start Envision server as follows:
    ```
    $ cd <path>/SMARTS/competition/track1/train
    $ scl envision start -s /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/
    ```
3. Open a new different second terminal and run the example as follows:
    ```
    $ cd <path>/SMARTS/competition/track1/train
    $ python3.8 run.py --mode=evaluate --model=/home/z/zhan1130/SMARTS/competition/track1/submission/best_model.zip
    ```

**Comment by zyzhang1130** [comment]
> python3.8 run.py --mode=evaluate --model=/home/z/zhan1130/SMARTS/competition/track1/submission/best_model.zip

I did as per instructed, still got the same error:
<img width="640" alt="Screenshot 2022-09-26 at 7 03 39 PM" src="https://user-images.githubusercontent.com/36942574/192260788-958f6c85-1c7e-4886-b74c-b513350728da.png">

extact inputs and outputs:
LHS
```
(base) zhan1130@xgpb0:~$ conda activate smarts
(smarts) zhan1130@xgpb0:~$ cd /home/z/zhan1130/SMARTS/competition/track1/train
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ scl envision start -s /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/
```
RHS:
```
(base) zhan1130@xgpb0:~$ conda activate smarts
(smarts) zhan1130@xgpb0:~$ cd /home/z/zhan1130/SMARTS/competition/track1/train
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ python3.8 run.py --mode=evaluate --model=/home/z/zhan1130/SMARTS/competition/track1/submission/best_model.zip

Torch cuda is available:  True 


Logdir: /home/z/zhan1130/SMARTS/competition/track1/train/logs/2022_09_26_15_55_45 


Model: /home/z/zhan1130/SMARTS/competition/track1/submission/best_model.zip 

Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_c/. ...
/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/gym/logger.py:34: UserWarning: WARN: Box bound precision lowered by casting to float32
  warnings.warn(colorize("%s: %s" % ("WARN", msg % args), "yellow"))
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_c/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_t/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/intersection/1_to_2lane_left_turn_t/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/merge/3lane_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/merge/3lane_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_cruise_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_cruise_single_agent/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_overtake/. ...
Waiting on /home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/smarts/scenarios/straight/3lane_overtake/. ...

Evaluate policy.

== CURRENT SYSTEM INFO ==
OS: Linux-5.4.0-126-generic-x86_64-with-glibc2.17 #142-Ubuntu SMP Fri Aug 26 12:12:57 UTC 2022
Python: 3.8.13
Stable-Baselines3: 1.4.0
PyTorch: 1.12.1+cu113
GPU Enabled: True
Numpy: 1.21.5
Gym: 0.19.0

== SAVED MODEL SYSTEM INFO ==
OS: Linux-5.13.0-44-generic-x86_64-with-glibc2.29 #49~20.04.1-Ubuntu SMP Wed May 18 18:44:28 UTC 2022
Python: 3.8.10
Stable-Baselines3: 1.4.0
PyTorch: 1.12.0+cu102
GPU Enabled: True
Numpy: 1.23.1
Gym: 0.19.0

/home/z/zhan1130/anaconda3/envs/smarts/lib/python3.8/site-packages/stable_baselines3/common/base_class.py:687: ResourceWarning: unclosed file <_io.BufferedReader name='/home/z/zhan1130/SMARTS/competition/track1/submission/best_model.zip'>
  data, params, pytorch_variables = load_from_zip_file(
ResourceWarning: Enable tracemalloc to get the object allocation traceback

Evaluating env 1_to_2lane_left_turn_c.
 Retrying in 0.05 seconds
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Evaluating env 1_to_2lane_left_turn_t.
 Retrying in 0.05 seconds
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Evaluating env 3lane_merge_single_agent.
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Evaluating env 3lane_cruise_single_agent.
 Retrying in 0.05 seconds
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Evaluating env 3lane_overtake.
 Retrying in 0.05 seconds
Could not connect to TraCI server at localhost:35473 [Errno 111] Connection refused
 Retrying in 0.05 seconds
Agent_0: Reached max episode steps.
Agent_0: Reached max episode steps.
Mean reward:0.00 +/- 0.00


Finished evaluating.

WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
WARNING:websocket:websocket connected
ERROR:websocket:'NoneType' object has no attribute 'sock' - goodbye
(smarts) zhan1130@xgpb0:~/SMARTS/competition/track1/train$ 
```
At this point I'm suspecting the security measure on the server side is preventing the Envision from executing successfully, since it potentially allows remote visual rendering?

**Comment by Adaickalavan** [comment]
The last message above indicates that the example code runs to completion and trains a model successfully. 

However, without access to the remote server, without access to the display (if any), without docker, and without sudo, I am afraid that we are unable to extend further assistance here to deduce issues with Envision visualization which might be related to the use of the remote server. 

Therefore, this issue is being closed. Please feel free to open a new issue for other questions.
